[processor/spanmetrics] Prune histograms

nijave opened the pull request:
Prune histograms when the dimension cache evictions are removed

**Description:**
Prunes histograms when the dimension cache is pruned. This prevents metric series from growing indefinitely

**Link to tracking Issue:**
 #27080

**Testing:**
I modified the the existing test to check `histograms` length instead of dimensions cache length. This required simulating ticks to hit the exportMetrics function

**Documentation:** <Describe the documentation added.>

nijave committed:
[processor/spanmetrics] Prune histograms

Prune histograms when the dimension cache evictions are removed

albertteoh reviewed (approved):
Thanks @nijave! 

Please be advised though that the [spanmetricsprocessor is deprecated](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/spanmetricsprocessor/README.md#span-metrics-processor). 

I believe the spanmetricsconnector does [prune histograms](https://github.com/albertteoh/opentelemetry-collector-contrib/blob/1005a5caea07d8d39c0ca95eb6a5f44e6334c165/connector/spanmetricsconnector/connector.go#L267), however, please let me know if I've misunderstood something.

mfilipe commented:
> Thanks @nijave!
> 
> Please be advised though that the [spanmetricsprocessor is deprecated](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/spanmetricsprocessor/README.md#span-metrics-processor).
> 
> I believe the spanmetricsconnector does [prune histograms](https://github.com/albertteoh/opentelemetry-collector-contrib/blob/1005a5caea07d8d39c0ca95eb6a5f44e6334c165/connector/spanmetricsconnector/connector.go#L267), however, please let me know if I've misunderstood something.

@albertteoh [I don't think it is being pruned](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27080#issuecomment-1734450345).

MovieStoreGuy commented:
Ahh, if this is related to memory management and concerns, then performing a `delete` on a map doesn't release the underlying storage until map is scheduled for GC. 

The easiest way to address this is use pointer values for the value type since it keeps the memory allocated small.

lastly, this is a deprecated component you're trying to update. Could you please apply the fix to the connector version instead?

nijave commented:
> Ahh, if this is related to memory management and concerns, then performing a `delete` on a map doesn't release the underlying storage until map is scheduled for GC.
> 
> The easiest way to address this is use pointer values for the value type since it keeps the memory allocated small.
> 
> lastly, this is a deprecated component you're trying to update. Could you please apply the fix to the connector version instead?

It's less about memory management and more about controlling Prometheus `active series` which get emitted to long term storage. We're seeing about an extra ~1m active series pile up which is $8k/mon on Grafana Cloud. Actually, the original workaround was to set Grafana Agent to have a 2GiB memory limit on Kubernetes and let it get OOMKill'd before it could pile up too many metrics.

Usually what's happening for us, some errors come in and generate new series. They drop off but the series stick around until the agents are restarted which can be weeks or even months. We also have batch jobs and other periodic workloads that emit trace data occasionally but not regularly and these pile up as well.

We're using Grafana Agent which pulls this in as a library and we need to rewrite/test our config using their new config pattern if we want to use connector so, for us, we need to update our deprecated Grafana Agent config first.

If I have time, I'll take a look at the connector and see if it has the same problem. I'm actually switching jobs so after Oct 6th, I won't have a real environment available to test changes anymore.

nijave commented:
> Ahh, if this is related to memory management and concerns, then performing a `delete` on a map doesn't release the underlying storage until map is scheduled for GC.

That is interesting--I'm not very well versed in `go` (especially the GC quirks) so I'll have to do some reading up but I suspect that's not an issue for the use case I described above.

MovieStoreGuy commented:
@nijave I've rerun the tests, could I ask you to address any failures and provide a changelog? 

Considering this is related to the Grafana, I'll cc @jpkrohling to see if they can also address this upstream as well :)

nijave commented:
> @nijave I've rerun the tests, could I ask you to address any failures and provide a changelog?
> 
> Considering this is related to the Grafana, I'll cc @jpkrohling to see if they can also address this upstream as well :)

Ok, I think I addressed all the issues. 

I have https://github.com/grafana/agent/issues/5271 open on Grafana Agent (I think it's linked in the corresponding issue in this repo)

MovieStoreGuy committed:
Merge branch 'main' into nv-prune-spanmetrics-histograms

nijave commented on .chloggen/27083-spanmetrics-prune.yaml:
```diff
@@ -0,0 +1,29 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
```
Not really sure about this one. Technically this will produce less metric series than it previously did but my understanding was it was always intended that metric series (histograms) get pruned at the same time as dimensions cache map.

i.e. it was a bug these histograms weren't being pruned
